### PR TITLE
check unused parameters

### PR DESCRIPTION
**Pull request by hnyu**
Add a code block for checking unused parameters, to help find potential bugs. This is also related to multi-gpu training

https://github.com/HorizonRobotics/alf/blob/pytorch/alf/utils/distributed.py#L104

### CODE REVIEW

File: alf/algorithms/algorithm.py [hunk 1]
```diff
@@ -1179,6 +1179,13 @@ def _backward_and_gradient_update(self, loss):
         if self._grad_scaler is not None:
             self._grad_scaler.update()
 
+        unused_parameters = []
```
Comment (breakds) [inline alf/algorithms/algorithm.py]: There are cases when this is actually desired. Can we make it optional so that the warning can also state "if this is intended, please add such config to silence it"?

Comment (hnyu) [reply alf/algorithms/algorithm.py]: > 

The warning will only appear once in the beginning, so I think it's fine?

Comment (breakds) [reply alf/algorithms/algorithm.py]: Sure. Which also makes the checking after the first warning a bit wasteful?

**Review by breakds — DISMISSED** [review]

File: alf/algorithms/algorithm.py (hunk 1 above)
Comment (hnyu) [reply alf/algorithms/algorithm.py]: > Sure. Which also makes the checking after the first warning a bit wasteful?

It's just a for loop which has been everywhere in this function. I'll merge it with the line below.

File: alf/algorithms/algorithm.py [hunk 2]
```diff
@@ -1180,6 +1180,9 @@ def _backward_and_gradient_update(self, loss):
             self._grad_scaler.update()
 
         all_params = [(self._param_to_name[p], p) for p in all_params]
+        unused_parameters = [p[0] for p in all_params if p[1].grad is None]
+        common.warning_once("Find unused parameters, please double check: %s",
```
Comment (emailweixu) [inline alf/algorithms/algorithm.py]: To be more specific: "Find parameters without gradients"

**Review by breakds — APPROVED** [review]
